src/api_server/src/lib.rs

```diff
@@ -60,6 +66,49 @@ impl fmt::Debug for Error {
 
 type Result<T> = std::result::Result<T, Error>;
 
+/// Repeatedly tries to connect to unix socket specified by `socket_path`.
+///
+/// Retries to connect `retries` times, with `interval` duration between tries.
+pub fn try_connect(
+    socket_path: PathBuf,
+    mut retries: u32,
+    interval: Duration,
+) -> io::Result<UnixStream> {
+    let mut sock = UnixStream::connect(socket_path.clone());
+    while retries > 0 {
+        if sock.is_ok() {
+            break;
+        }
+        std::thread::sleep(interval);
+        retries = retries - 1;
+        sock = UnixStream::connect(socket_path.clone());
+    }
+    sock
+}
+
+/// An owned handle to an ApiServer thread, used to stop and join the thread.
+pub struct ThreadedApiServerHandle {
+    stop_signum: c_int,
+    stop_flag: Arc<AtomicBool>,
+    thread_handle: JoinHandle<()>,
+}
+
+impl ThreadedApiServerHandle {
```

Comment:
nitpick: If another revision of this PR is made, consider renaming this type from `ThreadedApiServerHandle` to `ApiServerThreadHandle`. The former implies it is a handle to the whole server (all the threads included). Really a nitpick though.

Reply:
sure, can do once reviews are done; also feel free to just push this on top of the PR branch yourself.